packages/transformers/webextension/src/WebExtensionTransformer.js

```diff
@@ -267,30 +321,44 @@ function cspPatchHMR(policy: ?string) {
 export default (new Transformer({
   async transform({asset, options}) {
     const code = await asset.getCode();
-    const parsed = jsm.parse(code);
-    const data: any = parsed.data;
-    validateSchema.diagnostic(
-      WebExtensionSchema,
-      {
-        data: data,
-        source: code,
-        filePath: asset.filePath,
-      },
-      '@parcel/transformer-webextension',
-      'Invalid Web Extension manifest',
-    );
-    await collectDependencies(
-      asset,
-      data,
-      parsed.pointers,
-      Boolean(options.hmrOptions),
-    );
-    if (options.hmrOptions) {
-      // To enable HMR, we must override the CSP to allow 'unsafe-eval'
-      data.content_security_policy = cspPatchHMR(data.content_security_policy);
+    if (asset.pipeline == 'mv3-bg-sw') {
```

Comment:
The "proper" solution for this would be a runtime. `@parcel/runtime-hmr` achieves exactly this.

Reply:
Should this be moved into the HMR runtime? If so how would we differentiate between a background script and a normal script in the runtime (to figure out what to inject)?

Reply:
Maybe create another specific runtime plugin for this? And make it add the extra entry only if the context is service-worker (or something like that, you need to prevent it from being added to every JS bundle).

Reply:
Makes sense. This would also solve the problem of getting HMR working with a background page instead of background scripts in MV2 (which doesn't work right now). We'd need a way to inject some JS into an HTML file, would that also be possible from the same runtime?

Reply:
You can add additional JS entry assets to JS bundles, not to HTML.
But we already have some logic to ensure that in development, every HTML asset has at least one script tag so that HMR works:
https://github.com/parcel-bundler/parcel/blob/c8a64c9a533072c12681ca0805d1e4d7481ee53d/packages/transformers/html/src/HTMLTransformer.js#L47-L48

So it should be possible, you just have to figure out which bundle in the bundlegraph you want to add the asset to.

Reply:
Switching to a runtime might take a bit, and I won't have much time to work on it for the next month so it might be best to make that change in a separate PR and just have the HMR issues for `background.page` and the suboptimal architecture be known limitations for now.

Reply:
I've finished the switch to a runtime system now, this logic is removed. Hopefully this should work better.